src/services/upower/device.hpp

```diff
@@ -0,0 +1,174 @@
+#include "dbus_device.h"
+
+namespace qs::service::upower {
+
+class UPowerDeviceState: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		Charging = 1,
+		Discharging = 2,
+		Empty = 3,
+		FullyCharged = 4,
+		PendingCharge = 5,
+		PendingDischarge = 6,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceState::Enum status);
+};
+
+class UPowerDeviceType: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		LinePower = 1,
+		Battery = 2,
+		Ups = 3,
+		Monitor = 4,
+		Mouse = 5,
+		Keyboard = 6,
+		Pda = 7,
+		Phone = 8,
+		MediaPlayer = 9,
+		Tablet = 10,
+		Computer = 11,
+		GamingInput = 12,
+		Pen = 13,
+		Touchpad = 14,
+		Modem = 15,
+		Network = 16,
+		Headset = 17,
+		Speakers = 18,
+		Headphones = 19,
+		Video = 20,
+		OtherAudio = 21,
+		RemoteControl = 22,
+		Printer = 23,
+		Scanner = 24,
+		Camera = 25,
+		Wearable = 26,
+		Toy = 27,
+		BluetoothGeneric = 28,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceType::Enum type);
+};
+
+///! A device exposed through the UPower system service.
+class UPowerDevice: public QObject {
+	Q_OBJECT;
+	// clang-format off
+	/// The type of power device.
+	///
+	/// > [!INFO] When checking if the device is a battery, also verify that `powerSupply` is set to `true`.
+	Q_PROPERTY(UPowerDeviceType::Enum type READ type NOTIFY typeChanged);
+	/// If the device is a power supply.
+	Q_PROPERTY(bool powerSupply READ powerSupply NOTIFY powerSupplyChanged);
+	/// Current energy level of the device in watt-hours.
+	Q_PROPERTY(qreal energy READ energy NOTIFY energyChanged);
+	/// Maximum energy capacity of the device in watt-hours
+	Q_PROPERTY(qreal energyFull READ energyFull NOTIFY energyFullChanged);
+	/// Rate of energy change in watts (positive when charging, negative when discharging).
+	Q_PROPERTY(qreal energyRate READ energyRate NOTIFY energyRateChanged);
+	/// Estimated time until the device is fully discharged, in seconds.
+	///
+	/// Will be set to `0` if charging.
+	Q_PROPERTY(qreal timeToEmpty READ timeToEmpty NOTIFY timeToEmptyChanged);
+	/// Estimated time until the device is fully charged, in seconds.
+	///
+	/// Will be set to `0` if discharging.
+	Q_PROPERTY(qreal timeToFull READ timeToFull NOTIFY timeToFullChanged);
+	/// Current charge level as a percentage.
+	Q_PROPERTY(qreal percentage READ percentage NOTIFY percentageChanged);
+	/// If the device is currently present in the system.
+	Q_PROPERTY(bool isPresent READ isPresent NOTIFY isPresentChanged);
+	/// Current state of the device.
+	Q_PROPERTY(UPowerDeviceState::Enum state READ state NOTIFY stateChanged);
+	/// Capacity of the device as a percentage of its original capacity.
+	///
+	/// > [!WARNING] This property will be useless if the device comes from the system's [DisplayDevice](../upower).
+	Q_PROPERTY(qreal capacity READ capacity NOTIFY capacityChanged);
```

Comment:
Would this be for cases like battery degradation? Not sure capacity is the best name for that if so. Also clear up if `percentage` is `energy / energyFull` or `energy / energyFull * capacity`.

Additionally why the warning? Explain what conditions make the value useless so users can try to figure it out themselves on a non default device, and if possible use some non random value like `0` if the device does not support it, in addition to a `capacitySupported` property.

Reply:
I changed `capacity` to `healthPercentage` to make its behavior more clear. As for the warning, I don't think there is a *real* way to check if capacity it is supported so I can just set it to `false` for the DisplayDevice and `true` for any other device. The documentation for `percentage` has also been changed to clarify that it is equivalent to `energy / energyFull`.

Reply:
Looks like the property will be unset if it isn't supported, meaning you can check with `AbstractDBusProperty::exists()`.

Reply:
Didn't know this, thanks!

Reply:
Turns out the `exists()` method always comes out to true; even if it is a DisplayDevice which just returns `0`. My guess is UPower has all of its properties visible regardless of whether or not the device supports it.

The current way I implemented it is kinda arbitrary (by checking the device path) but that's all I can think of for implementing the `healthSupported` property - other than just checking if it is already `0`.

Reply:
I think its fair to say capacity isn't supported if its zero. better than checking if it's a display device at least.